onnxruntime/core/graph/contrib_ops/contrib_defs.cc

```diff
@@ -1923,6 +1923,248 @@ Matrix product that behaves like numpy.matmul: https://docs.scipy.org/doc/numpy-
+https://scikit-learn.org/stable/modules/generated/sklearn.preprocessing.OneHotEncoder.html
+`categories` attribute contains an array of category values.
+
+The input is a sparse tensor of values.
+
+The operator produces a sparse output with a dense shape that has one more dimension than input.
+That dimension contains sparsified version of 1-of-K category output.
+
+Example:
+categories = [1021, 2002]
+sparse input = [1021, 2002, 1023] dense shape: {10}
+input COO indices = [4, 7, 9]
+
+Output dense shape = {10, 2} // 2 is the number of categories, choices for the last dimension include
+{0, 0}, {1, 0}, {0, 1}
+
+{0, 0} is a combination that would either match no categories or it is absent from the sparse input.
+It will, therefore, be absent in the sparse output.
+
+For the rest of the values, we places ones in the following 2-D indices:
+(4, 0), (7, 1) and 1023 can not be categorized.
+
+The sparse output will have ones at the following flat COO indices: 8 (4 x 2) and 15 (7 x 2 + 1)
+)DOC";
+
+  ONNX_CONTRIB_OPERATOR_SCHEMA(OneHotEncoder)
+        .SetDomain(kMSDomain)
+        .SinceVersion(1)
+        .Input(0, "A", "2-dimensional sparse matrix A", "T")
+        .Output(0, "Y", "Sparsified 1-of-K output", "T1")
+        .Attr(
+            "categories",
+            "list of cat_ints",
+            AttributeProto::INTS)
+        .Attr(
+            "categories",
+            "list of cat_ints",
+            AttributeProto::STRINGS)
+        .TypeConstraint(
+            "T",
+            {"sparse_tensor(float)", "sparse_tensor(double)",
+             "sparse_tensor(int64)", "sparse_tensor(int32)",
+             "sparse_tensor(uint64)", "sparse_tensor(uint32)"},
+            "Constrain input and output types to numeric tensors.")
+        .TypeConstraint("T1", {"sparse_tensor(int64)"}, "index type")
+      .TypeAndShapeInferenceFunction([](ONNX_NAMESPACE::InferenceContext& ctx) {
+          propagateElemTypeFromDtypeToOutput(ctx, TensorProto_DataType_INT64, 0);
+          if (!hasInputShape(ctx, 0)) {
+            return;
+          }
+          int cat_len = 0;
+          const auto* cat_attr = ctx.getAttribute("cat_str");
+          if (cat_attr && cat_attr->type() == AttributeProto_AttributeType_STRINGS) {
+            cat_len = cat_attr->strings_size();
+          } else if ((cat_attr = ctx.getAttribute("cat_ints")) != nullptr &&
+                                cat_attr->type() == AttributeProto_AttributeType_INTS) {
+            cat_len = cat_attr->ints_size();
+          } else {
+            fail_shape_inference("Can't get categories attribute");
+          }
+          const auto& shape0 = getInputShape(ctx, 0);
+          TensorShapeProto output_shape;
+          *output_shape.add_dim() = shape0.dim(0);
+          output_shape.add_dim()->set_dim_value(cat_len);
+          updateOutputShape(ctx, 0, output_shape, TypeProto::kSparseTensorType);
+    });
+
   ONNX_CONTRIB_OPERATOR_SCHEMA(SparseToDenseMatMul)
```

Comment:
I missed DenseToSparseMatMul. We need that one too. In most ml pipeline, OneHotEncoder takes a dense tensor as input and outputs a sparse tensor. As your spec for OneHotEncoder only considers sparse inputs, we'll need an operator to convert dense tensor into sparse.